File: tests/go/test_stub.go
package go_test

import (
	"context"
	"testing"

	v1 "github.com/wjb-dev/demo-go-server/proto/v1"
)

type stubServer struct {
	v1.UnimplementedEchoServiceServer
}

func (s *stubServer) Echo(ctx context.Context, req *v1.EchoRequest) (*v1.EchoResponse, error) {
	return &v1.EchoResponse{Message: req.Message}, nil
}

func TestEcho(t *testing.T) {
	srv := &stubServer{}
	req := &v1.EchoRequest{Message: "hello"}
	resp, err := srv.Echo(context.Background(), req)
	if err != nil {
		t.Fatalf("Echo returned error: %v", err)
	}
	if resp.Message != "hello" {
		t.Errorf("expected %q, got %q", "hello", resp.Message)
	}
}

File: tests/cpp/test_stub.cc
#include <gtest/gtest.h>
#include "generated/service.pb.h"
#include "generated/service.grpc.pb.h"

using v1::EchoRequest;
using v1::EchoResponse;
using v1::EchoService;

class StubService : public EchoService::Service {
 public:
  grpc::Status Echo(grpc::ServerContext*, const EchoRequest* req,
                    EchoResponse* resp) override {
    resp->set_message(req->message());
    return grpc::Status::OK;
  }
};

TEST(EchoTest, EchoReturnsSameMessage) {
  StubService stub;
  EchoRequest req;
  req.set_message("hello");
  EchoResponse resp;
  grpc::ServerContext ctx;

  grpc::Status status = stub.Echo(&ctx, &req, &resp);
  EXPECT_TRUE(status.ok());
  EXPECT_EQ(resp.message(), "hello");
}
